infrastructure/local-provisioners/azurerm_apim.ts

```diff
@@ -0,0 +1,510 @@
+
+import apiManagementClient = require("azure-arm-apimanagement");
+import webSiteManagementClient = require("azure-arm-website");
+
+import * as fs from "fs";
+import * as path from "path";
+import * as shelljs from "shelljs";
+import * as tmp from "tmp";
+import * as url from "url";
+
+import { left } from "fp-ts/lib/Either";
+import * as replaceInFiles from "replace-in-file";
+import {
+  getObjectFromJson,
+  IResourcesConfiguration,
+  readConfig
+} from "../../lib/config";
+
+// tslint:disable-next-line:no-object-mutation
+shelljs.config.fatal = true;
+// shelljs.config.verbose = true;
+
+const CONFIGURATION_DIRECTORY_NAME = "apim";
+const CONFIGURATION_DIRECTORY_PATH = path.resolve(
+  __dirname,
+  `../${CONFIGURATION_DIRECTORY_NAME}`
+);
+
+/**
+ * ApiParams contains the variables taken from command line (provisioner arguments)
+ * that represents values that *change* between different deploying environments.
+ */
+const SetupOpenApiParams = t.interface({
+  environment: t.string,
+  apim_include_products: t.boolean,
+  apim_include_policies: t.boolean,
+  apim_task: t.literal("setupOpenapi")
+});
+
+const CreateApiManagementServiceParams = t.interface({
+  environment: t.string,
+  azurerm_apim_eventhub_connstr: t.string,
+  adb2c_tenant_id: t.string,
+  adb2c_portal_client_id: t.string,
+  adb2c_portal_client_secret: t.string,
+  apim_task: t.literal("createApiManagementService")
+});
+
+type ApimParams = t.TypeOf<typeof ApimParams>;
+
+const ApimParams = t.union([
+  SetupOpenApiParams,
+  CreateApiManagementServiceParams
+]);
+
+const addDays = (date: Date, days: number) => {
+  const result = new Date(date);
+  result.setDate(result.getDate() + days);
+  return result;
+};
+
+const setupOpenapi = (
+  apiClient: apiManagementClient,
+  config: IResourcesConfiguration,
+  backendUrl: string,
+  masterKey: string,
+  includeProducts: boolean,
+  includePolicies: boolean
+) => {
+  return Promise.all(
+    config.apim_apis.map(async apiEntry => {
+      const contentValue = `${backendUrl}${apiEntry.api.specsPath}?code=${
+        masterKey
+      }`;
+
+      winston.info(
+        `Adding API from URL: ${backendUrl}${apiEntry.api.specsPath}`
+      );
+
+      // Add API to API management
+      await apiClient.api.createOrUpdate(
+        config.azurerm_resource_group,
+        config.azurerm_apim,
+        apiEntry.id,
+        {
+          contentFormat: "swagger-link-json",
+          contentValue,
+          displayName: apiEntry.api.displayName,
+          path: apiEntry.api.path,
+          protocols: ["https"],
+          // WARNING: serviceUrl is taken from the swagger specs "host" field
+          // and there's no way to override that value here: it *must* be changed
+          // manually in the API management settings
+          // (or provide a real value in the swagger specs).
+          serviceUrl: backendUrl
+        }
+      );
+      // Add API to products
+      if (includeProducts) {
```

Comment:
would be useful to log the event of `includeProducts == false` in the `else` block, just to make the user aware of what's going on